File: frontend/src/api/transfer.ts
import api from './client'
import { toDownloadPayload, type DownloadResponsePayload } from './download'

export interface TransferDetectResponse {
  success: boolean
  message: string
  varieties: string[]
  count: number
}

export interface TransferDedupResponse {
  success: boolean
  message: string
  deduplicated: string[]
  removed_count: number
}

export interface TransferTemplateInfo {
  configured: boolean
  path: string
  filename: string
  updated_at: string
}

export interface TransferTemplateStatusResponse {
  templates: Record<string, TransferTemplateInfo>
}

export interface MonthlyTransferGroup {
  date: string
  files: string[]
  count: number
}

export interface MonthlyTransferPreviewResponse {
  success: boolean
  groups: MonthlyTransferGroup[]
  unrecognized_files: string[]
  total_files: number
  message: string
}

/**
 * Detect varieties contained in a batch of "big table" Excel files.  The
 * server scans the upload payloads and returns a deduplicated, sorted list of
 * variety names so the user can confirm before kicking off the migration.
 */
export function extractVarietiesFromUploads(files: File[] | FileList) {
  const formData = new FormData()
  Array.from(files).forEach((file) => {
    formData.append('files', file)
  })
  return api.post<TransferDetectResponse>('/api/transfer/varieties/upload', formData)
}

/**
 * Server-side deduplication for a comma/whitespace-separated list of veg
 * names.  Used by the transfer page so the user can paste a long list and
 * collapse duplicates before submitting.
 */
export function dedupVegNames(names: string[]) {
  return api.post<TransferDedupResponse>('/api/transfer/dedup', { veg_names: names })
}

/**
 * Inspect the persisted small-table templates that the backend keeps on the
 * filesystem.  Returns one entry per supported small-type kind.
 */
export function getTransferTemplates() {
  return api.get<TransferTemplateStatusResponse>('/api/transfer/templates')
}

/**
 * Upload a small-table template for one of the supported kinds (滨鲜 / 1号 / ...).
 */
export function uploadTransferTemplateFile(kind: string, file: File) {
  const formData = new FormData()
  formData.append('kind', kind)
  formData.append('template_file', file)
  return api.post<TransferTemplateStatusResponse>('/api/transfer/templates/upload', formData)
}

/**
 * Group a folder of "monthly" Excel uploads by date so the user can preview
 * which files will be processed before launching the long-running migration.
 */
export function previewMonthlyTransferUpload(params: { files: File[]; month: string }) {
  const formData = new FormData()
  params.files.forEach((file) => {
    formData.append('files', file)
  })
  formData.append('month', params.month)
  return api.post<MonthlyTransferPreviewResponse>('/api/transfer/monthly/preview', formData)
}

/**
 * Run the full transfer workflow against a batch of uploads.  The backend
 * streams the result file back and includes counters in the response headers.
 */
export async function executeTransferUpload(params: {
  tableFiles: File[]
  smallTemplate: File | null
  vegNames: string[]
  smallType: string
}): Promise<DownloadResponsePayload> {
  const formData = new FormData()
  params.tableFiles.forEach((file) => {
    formData.append('table_files', file)
  })
  if (params.smallTemplate) {
    formData.append('small_template_file', params.smallTemplate)
  }
  formData.append('veg_names_json', JSON.stringify(params.vegNames))
  formData.append('small_type', params.smallType)
  const response = await api.post('/api/transfer/execute/upload', formData, {
    responseType: 'blob',
  })
  return toDownloadPayload(response, 'transfer-result.xlsx')
}

/**
 * Run the monthly transfer workflow.  The server zips up the per-day results
 * and streams a single download back.
 */
export async function executeMonthlyTransferUpload(params: {
  files: File[]
  month: string
  smallType: string
  smallTemplate: File | null
}): Promise<DownloadResponsePayload> {
  const formData = new FormData()
  params.files.forEach((file) => {
    formData.append('files', file)
  })
  formData.append('month', params.month)
  formData.append('small_type', params.smallType)
  if (params.smallTemplate) {
    formData.append('small_template_file', params.smallTemplate)
  }
  const response = await api.post('/api/transfer/monthly/execute', formData, {
    responseType: 'blob',
  })
  return toDownloadPayload(response, `monthly-transfer-${params.month}.zip`)
}

/**
 * Browse a server-side directory.  Used by the directory picker dialog on the
 * data transfer page.
 */
export function browseTransferDirectory(path: string) {
  return api.get<{
    path: string
    parent: string | null
    items: { name: string; path: string; is_dir: boolean }[]
  }>('/api/transfer/browse', { params: { path } })
}

export interface BrowseDirectoryResponse {
  path: string
  subdirs: string[]
  files: string[]
}

export function browseDirectory(path: string) {
  return api.post<BrowseDirectoryResponse>('/api/transfer/browse', { path })
}

/**
 * Notify the backend when path-lock state is restored from localStorage cache,
 * so the server operator can see the event in the log / CMD window.
 */
export function logPathRestore(bigDir: string) {
  return api.post('/api/transfer/log-restore', { path: bigDir }).catch(() => {})
}

/**
 * Find .doc/.docx files in a server-side directory for path-locking mode.
 */
export function findTransferFiles(dir: string) {
  return api.post<BrowseDirectoryResponse>('/api/transfer/find-files', { path: dir })
}

export interface TransferVarietiesResponse {
  varieties: string[]
  count: number
}

/**
 * Extract varieties from server-side file paths (not uploads).
 */
export function extractVarietiesFromPaths(tablePaths: string[]) {
  return api.post<TransferVarietiesResponse>('/api/transfer/varieties', { table_paths: tablePaths })
}

/**
 * Execute a single transfer using server-side file paths, returns download blob.
 * When outputDir is provided the file is saved directly to that directory on the
 * server and a JSON response is returned instead of a blob.
 */
export async function executeTransferFromPaths(params: {
  tablePaths: string[]
  smallTemplatePath: string
  vegNames: string[]
  smallType: string
  outputDir?: string
}): Promise<DownloadResponsePayload> {
  const formData = new FormData()
  formData.append('table_paths_json', JSON.stringify(params.tablePaths))
  formData.append('small_template_path', params.smallTemplatePath)
  formData.append('veg_names_json', JSON.stringify(params.vegNames))
  formData.append('small_type', params.smallType)
  if (params.outputDir) {
    formData.append('output_dir', params.outputDir)
    const response = await api.post<{
      success: boolean
      message: string
      output_file: string
      processed_files: number
      matched_count: number
      written_count: number
    }>('/api/transfer/execute-from-paths', formData)
    return toDownloadPayload(response, 'transfer-result.docx')
  }
  const response = await api.post('/api/transfer/execute-from-paths', formData, {
    responseType: 'blob',
  })
  return toDownloadPayload(response, 'transfer-result.docx')
}

File: backend/api/routes/transfer.py
import asyncio
import json
import os
import re
import string
import tempfile
from datetime import datetime
from pathlib import Path

from fastapi import APIRouter, File, Form, HTTPException, UploadFile
from starlette.concurrency import run_in_threadpool

from backend.models.schemas import (
    BrowseRequest,
    BrowseResponse,
    DedupRequest,
    DedupResponse,
    DetectRequest,
    DetectResponse,
    TransferDetail,
    MonthlyTransferPreviewResponse,
    TransferRequest,
    TransferResponse,
    TransferTemplateStatusResponse,
    VarietiesRequest,
    VarietiesResponse,
)
from backend.api.upload_utils import (
    build_download_response,
    create_zip_archive,
    parse_json_form_value,
    save_upload,
    save_uploads,
)
from backend.services.doc_service import DocService
from backend.services.template_library_service import (
    get_transfer_template_path,
    get_transfer_templates,
    save_transfer_template,
)

import logging

_logger = logging.getLogger(__name__)

router = APIRouter()
doc_service = DocService()


def _run_in_executor(func, *args):
    loop = asyncio.get_event_loop()
    return loop.run_in_executor(None, func, *args)


@router.post("/detect", response_model=DetectResponse)
async def detect_tables(req: DetectRequest):
    if not os.path.isdir(req.folder_path):
        raise HTTPException(status_code=400, detail="文件夹路径无效")

    y, m, d = req.date.split("-")
    pattern = re.compile(
        rf"农残检测记录表{re.escape(y)}\.{re.escape(m)}\.{re.escape(d)}(?:-(\d+))?\.docx$"
    )

    detected = []
    for filename in os.listdir(req.folder_path):
        match = pattern.match(filename)
        if match:
            num = int(match.group(1)) if match.group(1) else 0
            detected.append((num, os.path.join(req.folder_path, filename)))

    detected.sort(key=lambda x: x[0])
    files = [path for _, path in detected]
    return DetectResponse(files=files, count=len(files))


@router.post("/varieties", response_model=VarietiesResponse)
async def extract_varieties(req: VarietiesRequest):
    if not req.table_paths:
        raise HTTPException(status_code=400, detail="未提供大表文件路径")

    varieties = doc_service.extract_all_varieties(req.table_paths)
    return VarietiesResponse(varieties=varieties, count=len(varieties))


@router.post("/varieties/upload", response_model=VarietiesResponse)
async def extract_varieties_from_uploads(
    table_files: list[UploadFile] = File(...),
):
    with tempfile.TemporaryDirectory(prefix="transfer-varieties-") as tmpdir:
        saved_paths = await save_uploads(table_files, Path(tmpdir) / "tables", "table")
        varieties = await run_in_threadpool(
            doc_service.extract_all_varieties,
            [str(path) for path in saved_paths],
        )
    return VarietiesResponse(varieties=varieties, count=len(varieties))


@router.post("/dedup", response_model=DedupResponse)
async def dedup_veg_names(req: DedupRequest):
    seen = set()
    deduped = []
    for name in req.veg_names:
        key = name.strip().lower()
        if key not in seen:
            seen.add(key)
            deduped.append(name)

    removed = len(req.veg_names) - len(deduped)
    return DedupResponse(
        original=req.veg_names,
        deduplicated=deduped,
        removed_count=removed,
    )


@router.get("/templates", response_model=TransferTemplateStatusResponse)
async def get_templates():
    return TransferTemplateStatusResponse(**get_transfer_templates())


@router.post("/templates/upload", response_model=TransferTemplateStatusResponse)
async def upload_template(
    small_type: str = Form(...),
    template_file: UploadFile = File(...),
):
    with tempfile.TemporaryDirectory(prefix="transfer-template-") as tmpdir:
        saved_path = await save_upload(template_file, Path(tmpdir), "small-template")
        try:
            status = save_transfer_template(small_type, saved_path, template_file.filename)
        except ValueError as exc:
            raise HTTPException(status_code=400, detail=str(exc)) from exc
    return TransferTemplateStatusResponse(**status)


@router.post("/monthly/preview", response_model=MonthlyTransferPreviewResponse)
async def preview_monthly_transfer(
    month: str = Form(...),
    table_files: list[UploadFile] = File(...),
):
    with tempfile.TemporaryDirectory(prefix="transfer-monthly-preview-") as tmpdir:
        saved_tables = await save_uploads(table_files, Path(tmpdir) / "tables", "table")
        result = doc_service.preview_monthly_groups([str(path) for path in saved_tables], month)
    return MonthlyTransferPreviewResponse(success=bool(result["groups"]), **result)


@router.post("/execute", response_model=TransferResponse)
async def execute_transfer(req: TransferRequest):
    if not req.table_paths:
        raise HTTPException(status_code=400, detail="未提供大表文件")
    if not req.small_template_path:
        raise HTTPException(status_code=400, detail="未提供小表模板")
    if not req.veg_names:
        raise HTTPException(status_code=400, detail="未提供菜名")
    if not req.output_dir:
        raise HTTPException(status_code=400, detail="未提供输出目录")

    timestamp = datetime.now().strftime("%Y%m%d_%H%M%S")
    output_path = os.path.join(req.output_dir, f"{req.small_type}_数据迁移_{timestamp}.docx")

    result = await _run_in_executor(
        doc_service.process_multiple_tables,
        req.table_paths,
        req.small_template_path,
        req.veg_names,
        output_path,
    )

    details = [
        TransferDetail(variety=d["variety"], rate=d["rate"], result=d["result"])
        for d in result.get("details", [])
    ]

    return TransferResponse(
        success=True,
        processed_files=result["processed_files"],
        matched_count=result["matched_count"],
        written_count=result["written_count"],
        output_file=result.get("output_file"),
        message=result["message"],
        details=details,
    )


@router.post("/execute/upload")
async def execute_transfer_upload(
    veg_names_json: str = Form(...),
    small_type: str = Form(default="small"),
    table_files: list[UploadFile] = File(...),
    small_template: UploadFile = File(...),
):
    veg_names = parse_json_form_value(
        veg_names_json,
        field_name="veg_names_json",
        expected_type=list,
    )
    normalized_veg_names = [
        str(item).strip()
        for item in veg_names
        if str(item).strip()
    ]
    if not normalized_veg_names:
        raise HTTPException(status_code=400, detail="未提供菜名")

    with tempfile.TemporaryDirectory(prefix="transfer-execute-") as tmpdir:
        root = Path(tmpdir)
        saved_tables = await save_uploads(table_files, root / "tables", "table")
        template_path = await save_upload(small_template, root / "template", "small-template")
        output_name = f"{small_type}_数据迁移_{datetime.now().strftime('%Y%m%d_%H%M%S')}.docx"
        output_path = root / "output" / output_name
        output_path.parent.mkdir(parents=True, exist_ok=True)

        result = await run_in_threadpool(
            doc_service.process_multiple_tables,
            [str(path) for path in saved_tables],
            str(template_path),
            normalized_veg_names,
            str(output_path),
        )

        generated_path = Path(result.get("output_file") or output_path)
        if not generated_path.exists():
            raise HTTPException(status_code=500, detail="数据迁移任务已结束，但没有生成输出文件")

        return build_download_response(
            generated_path,
            media_type="application/vnd.openxmlformats-officedocument.wordprocessingml.document",
            extra_headers={
                "X-Operation-Message": result.get("message", "数据迁移已完成"),
                "X-Processed-Files": str(result.get("processed_files", 0)),
                "X-Matched-Count": str(result.get("matched_count", 0)),
                "X-Written-Count": str(result.get("written_count", 0)),
            },
        )


@router.post("/monthly/execute")
async def execute_monthly_transfer_upload(
    month: str = Form(...),
    veg_names_json: str = Form(...),
    small_type: str = Form(default="滨鲜"),
    use_saved_template: str = Form(default="true"),
    table_files: list[UploadFile] = File(...),
    small_template: UploadFile | None = File(default=None),
):
    veg_names = parse_json_form_value(
        veg_names_json,
        field_name="veg_names_json",
        expected_type=list,
    )
    normalized_veg_names = [
        str(item).strip()
        for item in veg_names
        if str(item).strip()
    ]
    if not normalized_veg_names:
        raise HTTPException(status_code=400, detail="未提供菜名")

    with tempfile.TemporaryDirectory(prefix="transfer-monthly-") as tmpdir:
        root = Path(tmpdir)
        saved_tables = await save_uploads(table_files, root / "tables", "table")
        try:
            if small_template is not None and small_template.filename:
                template_path = await save_upload(small_template, root / "template", "small-template")
            else:
                template_path = get_transfer_template_path(small_type)
        except FileNotFoundError as exc:
            if use_saved_template.lower() == "false":
                raise HTTPException(status_code=400, detail="请上传本次使用的小表模板") from exc
            raise HTTPException(status_code=400, detail=str(exc)) from exc

        preview = doc_service.preview_monthly_groups([str(path) for path in saved_tables], month)
        if not preview["groups"]:
            raise HTTPException(status_code=400, detail="未识别到所选月份的大表文件")

        output_dir = root / "output"
        output_dir.mkdir(parents=True, exist_ok=True)
        manifest = {
            "month": month,
            "success": [],
            "skipped": [],
            "unrecognized_files": preview["unrecognized_files"],
            "generated_files": [],
        }

        for group in preview["groups"]:
            date_text = group["date"]
            output_name = f"{small_type}_数据迁移_{date_text.replace('-', '.')}.docx"
            output_path = output_dir / output_name
            result = await run_in_threadpool(
                doc_service.process_multiple_tables,
                group["files"],
                str(template_path),
                normalized_veg_names,
                str(output_path),
            )
            generated_path = Path(result.get("output_file") or output_path)
            if generated_path.exists():
                manifest["success"].append({
                    "date": date_text,
                    "input_files": [Path(path).name for path in group["files"]],
                    "matched_count": result.get("matched_count", 0),
                    "written_count": result.get("written_count", 0),
                    "output_file": generated_path.name,
                })
                manifest["generated_files"].append(generated_path.name)
            else:
                manifest["skipped"].append({
                    "date": date_text,
                    "input_files": [Path(path).name for path in group["files"]],
                    "reason": result.get("message", "未生成输出文件"),
                })

        manifest["success_count"] = len(manifest["success"])
        manifest["skipped_count"] = len(manifest["skipped"]) + len(manifest["unrecognized_files"])
        manifest["message"] = (
            f"月度数据迁移完成，成功 {manifest['success_count']} 天，"
            f"跳过 {manifest['skipped_count']} 项"
        )
        (output_dir / "处理结果清单.json").write_text(
            json.dumps(manifest, ensure_ascii=False, indent=2),
            encoding="utf-8",
        )

        generated_files = sorted(path for path in output_dir.iterdir() if path.is_file())
        generated_docx = [path for path in generated_files if path.suffix.lower() == ".docx"]
        if not generated_docx:
            detail = manifest["skipped"][0]["reason"] if manifest["skipped"] else "没有生成任何月度数据迁移文件"
            raise HTTPException(status_code=400, detail=detail)

        archive_path = create_zip_archive(root / f"数据迁移月度结果-{month}.zip", generated_files, arc_root=output_dir)
        return build_download_response(
            archive_path,
            media_type="application/zip",
            extra_headers={
                "X-Operation-Message": manifest["message"],
                "X-Processed-Files": str(sum(group["count"] for group in preview["groups"])),
                "X-Matched-Count": str(sum(item["matched_count"] for item in manifest["success"])),
                "X-Written-Count": str(sum(item["written_count"] for item in manifest["success"])),
                "X-Generated-Count": str(len(generated_docx)),
                "X-Skipped-Count": str(manifest["skipped_count"]),
            },
        )


@router.post("/find-files", response_model=BrowseResponse)
async def find_docx_files(req: BrowseRequest):
    """List .doc/.docx files in a server-side directory."""
    dir_path = (req.path or "").strip()
    if not dir_path or not os.path.isdir(dir_path):
        raise HTTPException(status_code=400, detail="目录路径无效")
    entries = os.listdir(dir_path)
    docx_files = sorted(
        e for e in entries
        if os.path.isfile(os.path.join(dir_path, e)) and e.lower().endswith((".doc", ".docx"))
    )
    return BrowseResponse(path=dir_path, subdirs=[], files=docx_files)


@router.post("/execute-from-paths")
async def execute_transfer_from_paths(
    table_paths_json: str = Form(...),
    small_template_path: str = Form(...),
    veg_names_json: str = Form(...),
    small_type: str = Form(default="small"),
    output_dir: str = Form(default=""),
):
    """Single transfer execution using server-side file paths.

    When ``output_dir`` is provided the output file is written directly to
    that directory and a JSON response is returned (no download).  Otherwise
    the legacy tempdir-and-download behaviour is preserved.
    """
    table_paths = parse_json_form_value(
        table_paths_json,
        field_name="table_paths_json",
        expected_type=list,
    )
    normalized_table_paths = [str(p).strip() for p in table_paths if str(p).strip()]
    if not normalized_table_paths:
        raise HTTPException(status_code=400, detail="未提供大表文件路径")

    veg_names = parse_json_form_value(
        veg_names_json,
        field_name="veg_names_json",
        expected_type=list,
    )
    normalized_veg_names = [
        str(item).strip()
        for item in veg_names
        if str(item).strip()
    ]
    if not normalized_veg_names:
        raise HTTPException(status_code=400, detail="未提供菜名")

    template_path = small_template_path.strip()
    if not os.path.exists(template_path):
        raise HTTPException(status_code=400, detail="小表模板路径无效")

    for p in normalized_table_paths:
        if not os.path.exists(p):
            raise HTTPException(status_code=400, detail=f"大表文件不存在: {p}")

    if output_dir.strip():
        target_dir = Path(output_dir.strip())
        if not target_dir.is_dir():
            raise HTTPException(status_code=400, detail="输出目录不存在")
        output_name = f"{small_type}_数据迁移_{datetime.now().strftime('%Y%m%d_%H%M%S')}.docx"
        output_path = target_dir / output_name
        output_path.parent.mkdir(parents=True, exist_ok=True)

        result = await run_in_threadpool(
            doc_service.process_multiple_tables,
            normalized_table_paths,
            template_path,
            normalized_veg_names,
            str(output_path),
        )

        generated_path = Path(result.get("output_file") or output_path)
        if not generated_path.exists():
            raise HTTPException(status_code=500, detail="数据迁移任务已结束，但没有生成输出文件")

        return {
            "success": True,
            "message": result.get("message", "数据迁移已完成"),
            "processed_files": result.get("processed_files", len(normalized_table_paths)),
            "matched_count": result.get("matched_count", 0),
            "written_count": result.get("written_count", 0),
            "output_file": str(generated_path),
        }

    # Legacy tempdir + download path
    with tempfile.TemporaryDirectory(prefix="transfer-execute-") as tmpdir:
        root = Path(tmpdir)
        output_name = f"{small_type}_数据迁移_{datetime.now().strftime('%Y%m%d_%H%M%S')}.docx"
        output_path = root / "output" / output_name
        output_path.parent.mkdir(parents=True, exist_ok=True)

        result = await run_in_threadpool(
            doc_service.process_multiple_tables,
            normalized_table_paths,
            template_path,
            normalized_veg_names,
            str(output_path),
        )

        generated_path = Path(result.get("output_file") or output_path)
        if not generated_path.exists():
            raise HTTPException(status_code=500, detail="数据迁移任务已结束，但没有生成输出文件")

        return build_download_response(
            generated_path,
            media_type="application/vnd.openxmlformats-officedocument.wordprocessingml.document",
            extra_headers={
                "X-Operation-Message": result.get("message", "数据迁移已完成"),
                "X-Processed-Files": str(result.get("processed_files", 0)),
                "X-Matched-Count": str(result.get("matched_count", 0)),
                "X-Written-Count": str(result.get("written_count", 0)),
            },
        )


@router.post("/log-restore")
async def log_path_restore(req: BrowseRequest):
    """Log when the frontend restores path-lock state from localStorage cache."""
    dir_path = (req.path or "").strip()
    _logger.info("数据迁移路径锁定状态已从缓存恢复，目录: %s", dir_path or "(空)")
    return {"ok": True}


@router.post("/browse", response_model=BrowseResponse)
async def browse_directory(req: BrowseRequest):
    path = (req.path or "").strip()

    if not path:
        if os.name == "nt":
            drives = [
                f"{letter}:\\"
                for letter in string.ascii_uppercase
                if os.path.exists(f"{letter}:\\")
            ]
            return BrowseResponse(path="", subdirs=drives, files=[])
        return BrowseResponse(path=os.path.sep, subdirs=[os.path.sep], files=[])

    if not os.path.isdir(path):
        raise HTTPException(status_code=400, detail="路径无效")

    entries = os.listdir(path)
    subdirs = sorted([e for e in entries if os.path.isdir(os.path.join(path, e))])
    files = sorted([e for e in entries if os.path.isfile(os.path.join(path, e))])
    return BrowseResponse(path=path, subdirs=subdirs, files=files)
